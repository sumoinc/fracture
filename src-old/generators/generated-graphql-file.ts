import { TypeScriptProject } from "projen/lib/typescript";
import { GeneratedFile, GeneratedFileOptions } from "./generated-file";

export type GeneratedGraphQlFileOptions = GeneratedFileOptions;

export class GeneratedGraphQlFile extends GeneratedFile {
  constructor(
    public readonly project: TypeScriptProject,
    filePath: string = "schema.graphql",
    options: GeneratedGraphQlFileOptions = {}
  ) {
    super(project, filePath, options);

    // ensure proper file exstension
    if (filePath.split(".").pop() !== "graphql") {
      throw new Error("GeneratedGraphQlFile must have a .graphql extension");
    }
  }

  preSynthesize() {
    this.addLine("# " + this.marker);
    this.addLine("\n");
    super.preSynthesize();
  }
}
