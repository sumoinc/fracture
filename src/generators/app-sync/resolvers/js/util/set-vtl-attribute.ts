import { ResourceAttributeGenerator } from "../../../../../services/resource-attribute";
import { StructureAttribute } from "../../../../../services/structure-attribute";
import { TypescriptStrategy } from "../../../../ts/strategy";

export const setVtlAttribute = (attribute: StructureAttribute) => {
  const strategy = TypescriptStrategy.of(attribute.project);
  const generator = attribute.generator;
  const name = strategy.formatAttributeName(attribute.name);
  const shortName = strategy.formatAttributeName(attribute.shortName);

  // non-generated
  if (generator === ResourceAttributeGenerator.NONE) {
    return `"${shortName}": $util.dynamodb.toDynamoDBJson($ctx.args.input.${name}),`;
    // generated values
  } else {
    switch (generator) {
      case ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP:
        return `"${shortName}": "$util.time.nowISO8601()"`;
      case ResourceAttributeGenerator.GUID:
        return `"${shortName}": "$util.autoId()"`;
      case ResourceAttributeGenerator.TYPE:
        return `"${shortName}": "unknown"`;

      default:
        throw new Error(`Unknown generator: ${generator}`);
    }
  }
};
