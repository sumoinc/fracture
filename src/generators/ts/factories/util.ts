import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
