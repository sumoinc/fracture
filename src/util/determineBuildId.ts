import { execSync } from "child_process";

/**
 * Returns the current git branch name, without folder.
 *
 * ie: feature/1234 returns 1234
 *
 * @returns {string}
 */
export const determineBuildId = () => {
  const branchName = execSync("git rev-parse --abbrev-ref HEAD")
    .toString("utf8")
    .replace(/[\n\r\s]+$/, "");
  const branchPath = branchName.split("/");
  return branchPath[branchPath.length - 1];
};
