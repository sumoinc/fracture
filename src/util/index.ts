export * from "./determineBuildId";
export * from "./test-util";
