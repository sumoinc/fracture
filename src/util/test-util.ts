import { Project } from "projen";
import { synthSnapshot } from "projen/lib/util/synth";
import { App } from "../apps";
import { FractureProject } from "../fracture-project";
import {
  DataService,
  Operation,
  Resource,
  ResourceAttribute,
  ResourceAttributeType,
  Structure,
} from "../services";
import { NuxtJsSite, VitePressSite } from "../sites";

export const TEST_ACCOUNT_ONE = "000000000000";
export const TEST_ORG_ONE = "org-123456";
export const TEST_REGION_ONE = "us-east-1";

/**
 *
 * Helper to generate and return one file from fracture.
 * Useful when trying to test the final output of generated files.
 *
 * @param fracture
 * @param filepath
 * @returns collection of files starting with supplied filepath
 */
export const synthFiles = (project: Project, filepath: string = ""): any => {
  const snapshot = synthSnapshot(project);
  // console.log(Object.keys(snapshot));
  const filtered = Object.keys(snapshot)
    .filter((path) => path.startsWith(filepath))
    .reduce((obj, key) => {
      return {
        ...obj,
        [key]: snapshot[key],
      };
    }, {} as { [key: string]: any });
  return filtered;
};

export const synthFile = (project: Project, filepath: string): string => {
  const files = synthFiles(project, filepath);

  if (files.length > 1) {
    throw new Error(`more than one file matched filepath "${filepath}"`);
  }

  if (files.length === 0) {
    throw new Error(`No files matched filepath "${filepath}"`);
  }

  return files[filepath];
};

/**
 * Builds a simple data service as a test harness
 */
export const testFractureProject = () => {
  return new FractureProject({
    name: "my-project",
  });
};

/**
 * Builds a simple data service as a test harness
 */
export const testApp = () => {
  return new App({
    parent: testFractureProject(),
    name: "my-app",
  });
};

/**
 * Builds a simple data service as a test harness
 */
export const testDataService = () => {
  return new DataService({
    parent: testFractureProject(),
    name: "my-service",
  });
};

/**
 * Build a simple resource to test with
 */
export const testResource = () => {
  return new Resource(testDataService(), { name: "my-resource" });
};

/**
 * A more complicated resource configuration
 */
export const complexService = () => {
  const service = testDataService();

  const person = new Resource(service, { name: "person" });

  const club = new Resource(service, {
    name: "club",
    comments: ["A club"],
    attributeOptions: [{ name: "name" }],
  });

  person.addAttribute({
    name: "first-name",
    shortName: "fn",
  });
  person.addAttribute({
    name: "last-name",
    shortName: "ln",
  });
  person.addAttribute({
    name: "height",
    shortName: "ht",
    type: ResourceAttributeType.INT,
  });
  person.addAttribute({
    name: "is-frog",
    shortName: "if",
    type: ResourceAttributeType.BOOLEAN,
    comments: ["Is this person a frog?"],
  });
  person.addAttribute({
    name: "fav-colors",
    type: ResourceAttributeType.ARRAY,
    typeParameter: ResourceAttributeType.STRING,
    comments: ["List opf favorite colors"],
  });

  // one specific club
  person.addAttribute({
    name: "one-club",
    type: club,
  });
  // an array of clubs
  person.addAttribute({
    name: "array-of-clubs",
    shortName: "ad",
    type: ResourceAttributeType.ARRAY,
    typeParameter: club,
  });

  // operation
  Operation.create(person);
  //Operation.read(person);
  //Operation.update(person);
  //Operation.delete(person);

  return service;
};

/**
 * Build a simple resource attribute to test with
 */
export const testResourceAttribute = () => {
  const resource = testResource();
  return new ResourceAttribute(resource.service, {
    name: "my-attr",
    resource,
  });
};

/**
 * Build a simple structure to test with
 */
export const testStructure = () => {
  const service = testDataService();
  return new Structure(service, { name: "my-structure" });
};

/**
 * Build a simple operation to test with
 */
export const testOperation = () => {
  const resource = testResource();
  return new Operation(resource.service, { name: "my-operation", resource });
};

/**
 * Build a nuxt site to test with
 */
export const testNuxtJsSite = () => {
  return new NuxtJsSite({
    parent: testFractureProject(),
    name: "my-site",
  });
};

/**
 * Build a vitepress site to test with
 */
export const testVitePressSite = () => {
  return new VitePressSite({
    parent: testFractureProject(),
    name: "my-site",
  });
};
