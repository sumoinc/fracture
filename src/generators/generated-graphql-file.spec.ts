import { GeneratedGraphQlFile } from "./generated-graphql-file";
import { synthFile, testDataService } from "../util/test-util";

describe("Success Conditions", () => {
  test("Smoke test - default name", () => {
    const service = testDataService();

    new GeneratedGraphQlFile(service);

    const content = synthFile(service, "src/schema.graphql");
    expect(content).toBeTruthy();
    expect(content).toMatchSnapshot();
    //console.log(content);
  });
  test("Smoke test - custom name", () => {
    const service = testDataService();

    new GeneratedGraphQlFile(service, "foo.graphql");

    const content = synthFile(service, "src/foo.graphql");
    expect(content).toBeTruthy();
    expect(content).toMatchSnapshot();
    // console.log(content);
  });
});

describe("Failure Conditions", () => {
  test("Wrong file extension", () => {
    expect(() => {
      new GeneratedGraphQlFile(testDataService(), "foo.ts");
    }).toThrow("extension");
  });
});
