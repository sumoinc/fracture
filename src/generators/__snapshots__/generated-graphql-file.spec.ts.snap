// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Success Conditions Smoke test - custom name 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

"
`;

exports[`Success Conditions Smoke test - default name 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

"
`;
